rest_framework_json_api/views.py

```diff
@@ -144,10 +144,15 @@ def retrieve_related(self, request, *args, **kwargs):
         if isinstance(instance, Iterable):
             serializer_kwargs['many'] = True
 
-        serializer = self.get_serializer(instance, **serializer_kwargs)
+        serializer = self.get_related_serializer(instance, **serializer_kwargs)
         return Response(serializer.data)
 
-    def get_serializer_class(self):
```

Comment:
removing this exposes the upstream DRF `get_serializer_class()`.  It seems this needs to remain for compatibility. @sliverc Do you agree?

Reply:
My whole point in making the change here is to do exactly this: expose the upstream method, so that it can be overwritten. `get_serializer_class()` is a public method on the base class. As such, it is part of DRF's public API - the [DRF documentation](https://www.django-rest-framework.org/api-guide/generic-views/#attributes) explicitly mentions that the method can be overwritten. In my opinion, an extension library should preserve the upstream API.

The current version of DJA does break the path to the upstream method. There might be cases where DJA users rely on this bug in some way. The change here would break their code. I'm just a first-time user of DJA without any idea about how it is being used in other projects, what the effects might be, and who might be affected.

Reply:
Forgive me for being dense but how is get_serializer_class() in DJA not able to be overridden by you? You've completely removed the function from DJA; not just added `get_related_serializer()`.

This extension library routinely overrides the upstream DRF because it is extending it.

Reply:
I do Full understand that you don't want to run the risk of breaking someone's application. And given that I am fairly new to DJA, I might not fully understand its inner workings, either.

I came across the issue the present PR is trying to solve, when I had to overwrite `get_serializer_class()` on a custom viewset, in exactly the same way I now added as part of the `AuthorViewSet` in the test example. I investigated the issue and discovered that when processing a GET request for a related instance, `RelatedMixin.get_related_instance()` directly accessed the `self.serializer_class` field instead of getting the class via its accessor method. Consequently, the upstream `get_serializer_class()` would not be called, either, which would be my overwritten version. Hence my remark that DJA effectively prevents `get_serializer_class()` to be overwritten on a custom viewset, which I consider to be a bug. 

My first attempt to fix the issue was to have `RelatedMixin.get_related_instance` call the upstream method instead of accessing the field directly. But this breaks things badly. During two exchanges with @sliverc (see the comments on the issue here) showed that the functionality of `RelatedMixin.get_serializer_class` is actually not about getting the serializer class, but getting the class of _related_ serializers. I followed the hint of @sliverc to _rename_ the method from `get_serializer_class()` to `get_related_serializer_class()`, which better captures its functionality. This way, `RelatedMixin.get_related_instance` can call `get_serializer_class()` of the parent class (potentially overwritten), while all functionality that has to do with getting serializers for related views is now handled by aptly named `get_related_*` methods.

To make this work, I need to do some other adjustments along the call chain. In particular, `retrieve_related()`, which is the entry point for a GET request on a related resource, now calls the renamed `get_related_serializer_class()`. One more change, also pointed out by @sliverc, was in `utils.get_resource_name()`: Here, I had to make the distinction if the  serializer of the present viewset is needed, or if the serializer of a related viewset should be returned. The distinction between the two cases depends on the `related_field` kwarg, similar to the condition in `get_serializer_class()`(not `get_related_serializer_class()`.

I hope this helps to provide some understanding of what I am trying to accomplish. I was very much focussed on the functionality of getting related resources; because of my limited understanding of DJA, I am not able to tell if other parts of its functionality are affected by the changes introduced in `RelatedMixin`. Here, I relied on the tests, which are all passing. If there is something else I can do to clear things up or to provide better explanation, please let me know. I am also available for a quick online call, if that can help to better assess the changes I introduced.

Reply:
Let me try to shed some light on this... `RelatedMixin` mistakenly shadowed `get_serializer_class` to get related serializer. Shadowing happened as `RelatedMixin` only works when used with a DRF generic view which provides a [get_serializer_class method](https://www.django-rest-framework.org/api-guide/generic-views/#get_serializer_classself) 
For `RelatedMixin` to be able to get parent serializer it then simply used `self.serializer_class` which is not DRF conform. However this only happened when [related urls](https://django-rest-framework-json-api.readthedocs.io/en/stable/usage.html#related-urls) have been configured as otherwise this [if](https://github.com/django-json-api/django-rest-framework-json-api/blob/master/rest_framework_json_api/views.py#L153) is always false and there is no difference than when the method hasn't been overwritten in `RelatedMixin`.

If someone overwrote `get_serializer_class` in its view without using related urls it would just work as `RelatedMixin` did not do anything anyway. If someone tried to overwrite `get_serializer_class` with related urls this user would run into exactly the same problem as @uliSchuster has.

Therefore I don't see that this change affects existing users.